### PR TITLE
bug fix:RevScan uses KeyMax to range descriptor lookup

### LINKED ISSUE #2298: RevScan [x,KeyMax) range descriptor lookup
I think when scanning a key range like the above in reverse order, we can run into the error
```go
return NewInvalidRangeMetaKeyError("body of meta2 range lookup is >= KeyMax", key)
```

in `ValidateRangeMetaKey`.
Logged here for investigation and writing a test.

### PR DESCRIPTION
**Pull request by thundercw**
Fixes #2298.

### CODE REVIEW

**Comment by thundercw** [comment]
@tschottdorf The bug is fixed according to your suggestion and some more tests in `replica_test.go`.
The code changed in  `replica_command.go` are most come from your branch 1392df2d68e7fd98480b5515ca08b505089058d8 .
PTAL.

File: keys/keys.go [hunk 1]
```diff
@@ -261,10 +261,10 @@ func ValidateRangeMetaKey(key proto.Key) error {
 	prefix, body := proto.Key(key[:len(Meta1Prefix)]), proto.Key(key[len(Meta1Prefix):])
 
 	if prefix.Equal(Meta2Prefix) {
```
Comment (tschottdorf) [inline keys/keys.go]: you're pretty much doing the same thing for Meta1 or Meta2 keys. Consider merging this here with the one below.

File: keys/keys.go [hunk 2]
```diff
@@ -246,9 +246,9 @@ func RangeMetaKey(key proto.Key) proto.Key {
 	return proto.KeyMin
 }
 
-// ValidateRangeMetaKey validates that the given key is a valid Range Metadata
+// validateRangeMetaKey validates that the given key is a valid Range Metadata
 // key.
```
Comment (tschottdorf) [inline keys/keys.go]: add to the comment: This checks only the constraints common to forward and backwards scans: correct prefix and not exceeding `KeyMax`.

File: keys/keys_test.go [hunk 3]
```diff
@@ -139,18 +139,28 @@ func TestMetaScanBounds(t *testing.T) {
 			expEnd:   Meta1Prefix.PrefixEnd(),
 		},
 		{
-			key:      proto.Key("foo"),
```
Comment (tschottdorf) [inline keys/keys_test.go]: what's this weird test case about?

File: keys/keys_test.go [hunk 4]
```diff
@@ -188,6 +198,11 @@ func TestMetaReverseScanBounds(t *testing.T) {
 			expStart: Meta1Prefix,
 			expEnd:   Meta2Prefix.Next(),
 		},
+		{
+			key:      Meta2KeyMax,
+			expStart: Meta2Prefix,
+			expEnd:   Meta2KeyMax.Next(),
+		},
```
Comment (tschottdorf) [inline keys/keys_test.go]: we should have some error cases here and in the forward scan tests as well. Could you add `expError string` to the struct and put in tests that fail specific to the scan direction (test the error with `IsError`).

File: storage/replica_command.go [hunk 5]
```diff
@@ -524,35 +520,68 @@ func (r *Replica) RangeLookup(batch engine.Engine, args proto.RangeLookupRequest
...
+			// An error here is likely a WriteIntentError when reading consistently.
 			return reply, nil, err
 		}
+	} else {
+		// Use MVCCScan to get first the first range. There are three cases:
+		// 1. args.Key is not an endpoint of the range and
+		// 2a. The args.Key is the start/end key of the range.
+		// 2b. Even worse, the body of args.Key is proto.KeyMax.
+		// In the first case, we need use the MVCCScan() to get the first range
+		// descriptor, because ReverseScan can't do the work.If we have ranges
```
Comment (tschottdorf) [inline storage/replica_command.go]: s/\.If/ If/

File: storage/replica_command.go [hunk 6]
```diff
@@ -524,35 +520,68 @@ func (r *Replica) RangeLookup(batch engine.Engine, args proto.RangeLookupRequest
...
+		// ["a","c") and ["c","f") and the reverse scan request's key range is
```
Comment (tschottdorf) [inline storage/replica_command.go]: I know we're being a little inconsistent, but I think the form `[a,b)` is sufficiently legible for comments and cuts some of the fat (i.e. not `["a","b")` or `[a, b)`).

File: storage/replica_command.go [hunk 7]
```diff
@@ -524,35 +520,68 @@ func (r *Replica) RangeLookup(batch engine.Engine, args proto.RangeLookupRequest
...
+		// ["b","d"), then "d".Next() is less than "f", and so the meta row
+		// {"f"->["c", "f")} would be ignored by MVCCReverseScan. In case 2a,
+		// the range descriptor received by MVCCScan will be filtered before
+		// results are returned: With ranges ["c", "f") and ["f","z"), reverse
+		// scan on ["d", "f") receives the descriptor {"z"->["f","z")}, which
+		// is discarded below since it's not being asked for. Finally, in case
+		// 2b, we don't even attempt the forward scan because it's neither
+		// defined nor required.
+		// TODO(tschottdorf): unit testing coverage.
+		if args.Key.Less(keys.Meta2KeyMax) {
```
Comment (tschottdorf) [inline storage/replica_command.go]: do we not have the same problem for `Meta1KeyMax`?

**Comment by tschottdorf** [comment]
Could you add some test cases to the `RangeLookup` tests in `client_range_test.go`? Exercising all of the reverse scan cases seems worthwhile (then my TODO in `RangeLookup` can be removed). In particular, scan and reverse scan with `KeyMax` on both meta levels, and with `StartKey` of a range (I think you already wrote those, but just check). Thanks!

File: keys/keys_test.go (hunk 3 above)
Comment (thundercw) [inline keys/keys_test.go]: Before putting `validateRangeMetaKey` into `MetaScanBounds`, the test can pass.

Comment (tschottdorf) [inline keys/keys_test.go]: :+1:

File: keys/keys_test.go (hunk 4 above)
Comment (thundercw) [inline keys/keys_test.go]: `testutils.IsError(err, test.expError)` can`t be used in pkg keys:
```
# github.com/cockroachdb/cockroach/keys
import cycle not allowed in test
package github.com/cockroachdb/cockroach/keys (test)
	imports github.com/cockroachdb/cockroach/testutils
	imports github.com/cockroachdb/cockroach/keys
```

File: storage/replica_command.go (hunk 7 above)
Comment (thundercw) [inline storage/replica_command.go]: Meta1KeyMax is legal for Scan. I add a test case in `client_range_test.go` about Meta1KeyMax, the result is what we want.

**Comment by thundercw** [comment]
@tschottdorf  I have add some test cases to the RangeLookup tests in client_range_test.go. Whether or not enough?

File: storage/replica_command.go (hunk 5 above)
Comment (thundercw) [inline storage/replica_command.go]: done

File: storage/replica_command.go (hunk 6 above)
Comment (thundercw) [inline storage/replica_command.go]: done

File: storage/replica_command.go [hunk 8]
```diff
@@ -605,45 +639,30 @@ func (r *Replica) RangeLookup(batch engine.Engine, args proto.RangeLookupRequest
...
+			}
+		}
 	}
 
-	// Decode all scanned range descriptors, stopping if a range is encountered
-	// which does not have the same metadata prefix as the queried key.
-	rds := make([]proto.RangeDescriptor, len(kvs))
-	for i := range kvs {
+	// Decode all scanned range descriptors which haven't been unmarshaled yet.
+	for i, l := len(rds), len(kvs); i < l; i++ {
```
Comment (tschottdorf) [inline storage/replica_command.go]: see https://github.com/cockroachdb/cockroach/pull/2340/files#r38758493

File: storage/client_range_test.go [hunk 9]
```diff
@@ -327,22 +327,53 @@ func TestRangeLookupUseReverse(t *testing.T) {
...
 	for _, split := range splits {
+		split.Txn = txn
 		_, err := store.ExecuteCmd(context.Background(), &split)
 		if err != nil {
 			t.Fatalf("%q: split unexpected error: %s", split.SplitKey, err)
 		}
 	}
 
+	// Resolve the intents.
+	rArgs := proto.ResolveIntentRangeRequest{
```
Comment (tschottdorf) [inline storage/client_range_test.go]: could you resolve the intents with a `Scan` (over the whole meta key range) and `SucceedsWithin` instead? That should be shorter and less confusing (admittedly, not faster).

File: keys/keys.go [hunk 10]
```diff
@@ -260,43 +261,49 @@ func ValidateRangeMetaKey(key proto.Key) error {
...
 
-	if prefix.Equal(Meta2Prefix) {
-		if body.Less(proto.KeyMax) {
-			return nil
-		}
-		return NewInvalidRangeMetaKeyError("body of meta2 range lookup is >= KeyMax", key)
-	}
-
-	if prefix.Equal(Meta1Prefix) {
+	if prefix.Equal(Meta2Prefix) || prefix.Equal(Meta1Prefix) {
```
Comment (tschottdorf) [inline keys/keys.go]: slight improvement:

```go
if !prefix.Equal(Meta2Prefix) && !prefix.Equal(Meta1Prefix) {
  return // not a meta key error
}
if proto.KeyMax.Less(body) {
  return // key > KeyMax error
}
return nil
```

File: keys/keys.go [hunk 11]
```diff
@@ -309,8 +316,8 @@ func MetaReverseScanBounds(key proto.Key) (proto.Key, proto.Key, error) {
 	// in the same prefix. For MVCCReverseScan the endKey is exclusive, if we want to find
 	// the range descriptor the given key specified,we need to set the key.Next() as the
 	// MVCCReverseScan`s endKey. For example:
-	// If we have ranges ["", "f") and ["f", "z"), then we'll have corresponding meta records
-	// at "f" and "z". If you're looking for the meta record for key "f", then you want the
+	// If we have ranges [a, f) and [f,z), then we'll have corresponding meta records
```
Comment (tschottdorf) [inline keys/keys.go]: s/ f/f/

File: storage/replica_command.go [hunk 12]
```diff
@@ -530,35 +526,68 @@ func (r *Replica) RangeLookup(batch engine.Engine, args proto.RangeLookupRequest
...
+		// range is [b,d), then d.Next() is less than "f", and so the meta
+		// row {f->[c,f)} would be ignored by MVCCReverseScan. In case 2a,
+		// the range descriptor received by MVCCScan will be filtered before
+		// results are returned: With ranges [c,f) and [f,z), reverse scan
+		// on [d,f) receives the descriptor {z->[f,z)}, which is discarded
+		// below since it's not being asked for. Finally, in case 2b, we
+		// don't even attempt the forward scan because it's neither defined
+		// nor required.
+		// TODO(tschottdorf): unit testing coverage.
+		if args.Key.Less(keys.Meta2KeyMax) {
```
Comment (tschottdorf) [inline storage/replica_command.go]: ping about hard-coding meta2. What about meta1?

**Comment by tschottdorf** [comment]
@thundercw: test cases look good. Had some other comments though, after that we're good to go.

File: storage/replica_command.go (hunk 12 above)
Comment (tschottdorf) [inline storage/replica_command.go]: sorry, I saw you responded already.

File: storage/replica_command.go (hunk 7 above)
Comment (tschottdorf) [inline storage/replica_command.go]: ok. Could you add something like:

// Note that Meta1KeyMax is admissible: it means we're looking for the range descriptor that houses `Meta2KeyMax`, and a forward scan handles it correctly.

File: keys/keys_test.go [hunk 13]
```diff
@@ -132,25 +133,65 @@ func TestMetaScanBounds(t *testing.T) {
...
+			expEnd:   nil,
+			expError: "body of meta key range lookup is",
 		},
 	}
 	for i, test := range testCases {
-		resStart, resEnd := MetaScanBounds(test.key)
+		resStart, resEnd, err := MetaScanBounds(test.key)
+
+		if err != nil {
+			matched, merr := regexp.MatchString(test.expError, err.Error())
```
Comment (tamird) [inline keys/keys_test.go]: you can use `testutils.IsError` now

Comment (thundercw) [inline keys/keys_test.go]: Cool.

File: storage/replica_command.go (hunk 8 above)
Comment (thundercw) [inline storage/replica_command.go]: done

File: storage/replica_command.go (hunk 7 above)
Comment (thundercw) [inline storage/replica_command.go]: done

File: storage/client_range_test.go (hunk 9 above)
Comment (thundercw) [inline storage/client_range_test.go]: done

File: keys/keys.go (hunk 10 above)
Comment (thundercw) [inline keys/keys.go]: done

File: keys/keys.go (hunk 11 above)
Comment (thundercw) [inline keys/keys.go]: done

**Comment by thundercw** [comment]
@tschottdorf @tamird  updated per your comments and squashed. PTAL.
Some test checks were not passed. From the results look less like my problem. I will rebase the code after #2367 be fixed and have a check.

**Comment by tamird** [comment]
#2367 is not what failed on your build. `TestRangeDescriptorSnapshotRace` failed on yours[0] with an NPE  though I haven't looked into more than that.

[0] https://circleci.com/gh/cockroachdb/cockroach/6983

File: keys/keys.go [hunk 14]
```diff
@@ -259,44 +260,49 @@ func ValidateRangeMetaKey(key proto.Key) error {
...
+	if proto.KeyMax.Less(body) {
+		return NewInvalidRangeMetaKeyError("body of meta key range lookup is > KeyMax", key)
 	}
-	return NewInvalidRangeMetaKeyError("not a meta key", key)
+	return nil
 }
 
 // MetaScanBounds returns the range [start,end) within which the desired meta
 // record can be found by means of an engine scan. The given key must be a
 // valid RangeMetaKey as defined by ValidateRangeMetaKey.
```
Comment (tamird) [inline keys/keys.go]: s/ValidateRangeMetaKey/validateRangeMetaKey/g

File: keys/keys_test.go [hunk 15]
```diff
@@ -132,25 +133,60 @@ func TestMetaScanBounds(t *testing.T) {
...
+			expEnd:   nil,
+			expError: "body of meta key range lookup is",
 		},
 	}
 	for i, test := range testCases {
-		resStart, resEnd := MetaScanBounds(test.key)
+		resStart, resEnd, err := MetaScanBounds(test.key)
+
+		if err != nil && !testutils.IsError(err, test.expError) {
+			t.Errorf("expected error: %s ; got %s", test.expError, err.Error())
```
Comment (tamird) [inline keys/keys_test.go]: s/.Error()/g

File: storage/client_range_test.go [hunk 16]
```diff
@@ -343,6 +341,36 @@ func TestRangeLookupUseReverse(t *testing.T) {
...
+	scanArgs := proto.ScanRequest{
+		RequestHeader: proto.RequestHeader{
+			Key:     proto.KeyMin,
+			EndKey:  keys.RangeMetaKey(proto.KeyMax),
+			RangeID: 1,
+			Replica: proto.Replica{StoreID: store.StoreID()},
+		},
+	}
+	util.SucceedsWithin(t, time.Second, func() error {
+		if _, err := store.ExecuteCmd(context.Background(), &scanArgs); err != nil {
```
Comment (tamird) [inline storage/client_range_test.go]: nit: remove this nil check.
```
_, err := store.ExecuteCmd(context.Background(), &scanArgs)
return err
```

File: storage/replica_command.go [hunk 17]
```diff
@@ -530,35 +526,70 @@ func (r *Replica) RangeLookup(batch engine.Engine, args proto.RangeLookupRequest
...
+				return reply, nil, err
+			}
 
+			kvs, intents, err = engine.MVCCScan(batch, startKey, endKey, 1,
+				args.Timestamp, consistent, args.Txn)
+			if err != nil {
+				return reply, nil, err
+			}
+			var firstRD proto.RangeDescriptor
+			if err = gogoproto.Unmarshal(kvs[0].Value.Bytes, &firstRD); err != nil {
```
Comment (tamird) [inline storage/replica_command.go]: s/=/:=/

File: keys/keys.go (hunk 14 above)
Comment (thundercw) [inline keys/keys.go]: done

File: keys/keys_test.go (hunk 15 above)
Comment (thundercw) [inline keys/keys_test.go]: done

File: storage/client_range_test.go (hunk 16 above)
Comment (thundercw) [inline storage/client_range_test.go]: done

File: storage/replica_command.go (hunk 17 above)
Comment (thundercw) [inline storage/replica_command.go]: done.  Why not use the previous？

Comment (tamird) [inline storage/replica_command.go]: this style is slightly more hygienic because it avoids changing the value of `err` in the scope.

Comment (thundercw) [inline storage/replica_command.go]: :+1:  Thank you for your reply.

**Comment by tschottdorf** [comment]
LGTM. Thanks @thundercw

**Comment by tschottdorf** [comment]
@thundercw: I'm going to merge this for you. Hope that's ok.

**Pull request merged**